src/main/PersistentState.cpp

```diff
@@ -4,32 +4,111 @@
 
 #include "PersistentState.h"
 
+#include "crypto/Hex.h"
 #include "database/Database.h"
 #include "herder/Herder.h"
+#include "herder/HerderUtils.h"
 #include "ledger/LedgerManager.h"
 #include "util/GlobalChecks.h"
 #include "util/Logging.h"
-#include "xdr/Stellar-internal.h"
 #include <Tracy.hpp>
 
 namespace stellar
 {
 
 using namespace std;
 
+std::chrono::minutes const PersistentState::TX_SET_GC_DELAY(1);
+
 std::string PersistentState::mapping[kLastEntry] = {
     "lastclosedledger", "historyarchivestate", "lastscpdata",
     "databaseschema",   "networkpassphrase",   "ledgerupgrades",
-    "rebuildledger",    "lastscpdataxdr"};
+    "rebuildledger",    "lastscpdataxdr",      "txset"};
 
 std::string PersistentState::kSQLCreateStatement =
     "CREATE TABLE IF NOT EXISTS storestate ("
     "statename   CHARACTER(32) PRIMARY KEY,"
     "state       TEXT"
     "); ";
 
-PersistentState::PersistentState(Application& app) : mApp(app)
+PersistentState::PersistentState(Application& app)
+    : mApp(app), mTxSetGarbageCollectTimer(app)
+{
+}
+
+void
+PersistentState::startTxSetGCTimer()
```

Comment:
I'm not a fan of the suggested approaches, because they both make it harder to reason about what's happening:
* we don't know how many times SCP persistence is called per ledger, so we can't really reason about how often tx sets are cleaned up (same thing with the number of tx sets).
* It makes testing harder as now we have to into account cases where purging triggers at different times depending on consensus.
* I went with the timer, as it seems to be the simplest approach that avoids anything complicated where we have to map Herder's state to the database and try to estimate a good time to do the cleanup.

I suggest we keep the timer approach. I do like the idea of moving it to Herder, this way we don't need extra start/shutdown logic in PersistentState, and it'll be easy to refactor the purging function based on your other suggestion.

Reply:
actually you're right, I was thinking that we could get arbitrary number of values from other validators, so we should try to GC more often in some situations, but this is only tracking what the current validator is doing.